web/modules/custom/portland/src/Plugin/Filter/FilterAccordionHeadings.php

```diff
@@ -0,0 +1,113 @@
+<?php
+
+namespace Drupal\portland\Plugin\Filter;
+
+use Drupal\Component\Utility\Html;
+use Drupal\filter\FilterProcessResult;
+use Drupal\filter\Plugin\FilterBase;
+
+/**
+ * Provides a filter to modify accordion headings.
+ *
+ * @Filter(
+ *   id = "filter_accordion_headings",
+ *   title = @Translation("Modify accordion headings to maintain heading hierarchy"),
+ *   description = @Translation("Modify accordion headings for accessibility. Must run after the CKEditor Accordion filter."),
+ *   type = Drupal\filter\Plugin\FilterInterface::TYPE_TRANSFORM_IRREVERSIBLE,
+ * )
+ */
+class FilterAccordionHeadings extends FilterBase
+{
+  /**
+   * {@inheritdoc}
+   */
+  public function process($text, $langcode)
+  {
+    $result = new FilterProcessResult($text);
+
+    $dom = Html::load($text);
+    $xpath = new \DOMXPath($dom);
+
+    // Find all headings h1 to h6 and store them in a flat array.
+    $headings = [];
```

Comment:
I think we could remove this for loop and the `isWIthinAccordionPanel` method, instead using the XPath query to make sure it's not a child of a panel:

```
// Find all headings h1 to h6 and store them in a flat array. Skips headings within accordion panels.
$headings = $xpath->query('(//h1 | //h2 | //h3 | //h4 | //h5 | //h6)[not(ancestor::*[@data-aria-accordion-panel])]');
```

I tested this and it appears to work the same and cleans up the code a bit.

Reply:
Nice! I'm not familiar with XPath. Glad to see the selector is so powerful. Fixed according to the feedback.